app/integration/runtime/src/main/java/io/syndesis/integration/runtime/handlers/TemplateStepHandler.java

```diff
@@ -0,0 +1,225 @@
+
+    @Override
+    public boolean canHandle(Step step) {
+        if (StepKind.template != step.getStepKind()) {
+            return false;
+        }
+
+        Action action = step.getAction().orElse(null);
+        if (action == null) {
+            return false;
+        }
+
+        DataShape inputDataShape = action.getInputDataShape().orElse(null);
+        if (inputDataShape == null) {
+            return false;
+        }
+
+        return DataShapeKinds.JSON_SCHEMA == inputDataShape.getKind();
+    }
+
+    @SuppressWarnings("PMD.AvoidReassigningParameters")
+    @Override
+    public Optional<ProcessorDefinition<?>> handle(Step step, ProcessorDefinition<?> route, IntegrationRouteBuilder builder, String flowIndex, String stepIndex) {
+        ObjectHelper.notNull(route, "route");
+
+        Map<String, String> properties = step.getConfiguredProperties();
+        String template = properties.get(TEMPLATE_PROPERTY);
+
+        /*
+         * To avoid creating a temporary file, we use the mustache template header
+         * to override the camel-mustache resource uri. This is more efficient but has
+         * a downside.
+         *
+         * Once returned, the route (ProcessorDefinition) is resolved. This resolution
+         * looks for any properties delimited by {{ and }}, which is the default syntax
+         * for mustache properties. Thus, resolution of the mustache properties is
+         * incorrectly attempted and unsurprisingly fails since these properties are
+         * meant for mustache and not camel.
+         *
+         * Changing the mustache delimiter patterns avoids this problem and allows the
+         * properties to be resolved later by mustache.
+         */
+        template = template.replaceAll(DOUBLE_OPEN_BRACE_PATTERN, MUSTACHE_OPEN_DELIMITER);
+        template = template.replaceAll(DOUBLE_CLOSE_BRACE_PATTERN, MUSTACHE_CLOSE_DELIMITER);
+
+        //
+        // Convert the exchange's in message from JSON
+        // to a HashMap since this is required by camel modules
+        //
+        route = route.process(jsonToMapProcessor);
+
+        //
+        // Apply the template to the mustache header property
+        // Then add to the route path
+        //
+        route.setHeader(MustacheConstants.MUSTACHE_TEMPLATE).constant(template);
+
+        //
+        // Encode the delimiters since they are applied as URI query parameters
+        //
+        try {
+            String openDelimiter = URLEncoder.encode(MUSTACHE_OPEN_DELIMITER, "UTF-8");
+            String closeDelimiter = URLEncoder.encode(MUSTACHE_CLOSE_DELIMITER, "UTF-8");
+
+            StringBuilder uri = new StringBuilder()
+                       .append("mustache")
+                       .append(COLON).append("dummy")
+                       .append(QUESTION_MARK).append("startDelimiter").append(EQUALS).append(openDelimiter)
+                       .append(AMPERSAND).append("endDelimiter").append(EQUALS).append(closeDelimiter);
```

Comment:
This may be problematic as if you add two templates in the same context, they will class, you should name the endpoint after the flow and step index.

I guess you can also use Camel's URISupport to create an uri.

Reply:
@lburgazzoli 
Yeah, I'll update the "dummy" literal to something unique as you suggest.
These template camel modules do support an URI but are limited to looking for the source of the URI as a file on the classpath (unless its an absolute file). Seeing as I already have the content and they support placing it in the header, it's become more straightforward to leave it that way.

Reply:
What I mean here is that instead of using StringBuilder, you can leverage Camel's `URISupport.appendParametersToURI(String, Map)` to generate the final uri. Not a blocking issue of course :)